contracts/Orchestrator.sol

```diff
@@ -0,0 +1,181 @@
+        distributionBackend = newBackend;
+    }
+
+    /// @notice Sets a new USDC pool address. Tokens from USDC pool
+    ///         will be used during orders execution.
+    /// @param pool The address of the pool
+    function setUsdcPool(address pool) public onlyOwner {
+        usdcPool = pool;
+    }
+
+    /// @notice Starts token distribution. This prohibits token
+    ///         unlocks and allows burning tokens from tender
+    ///         addresses.
+    function startDistribution() public onlyBackend {
+        blend.startDistributionPhase();
+    }
+
+    /// @notice Ends token distribution. This action reverts BLND
+    ///         to the regurlar phase, i.e. unlocks are allowed,
+    ///         and burns are prohibited.
+    function stopDistribution() public onlyBackend {
+        blend.stopDistributionPhase();
+    }
+
+    /// @notice Executes orders from lowest price to highest.
+    ///         If there is not enough funds on either BLND
+    ///         balance or USDC balance, order is executed
+    ///         partially. If the specified tender address
+    ///         does not exist, the order is skipped.
+    function executeOrders(Order[] memory orders) public onlyBackend {
+        require(
+            blend.distributionPhase(),
+            "Current phase does not allow distribution"
+        );
+        uint256 lastPrice = 0;
+        for (uint i = 0; i < orders.length; i++) {
+            require(orders[i].price >= lastPrice, "Orders must be sorted");
+            uint256 usdcLeft = usdc.balanceOf(usdcPool);
+            uint256 allowanceLeft = usdc.allowance(usdcPool, address(this));
+            uint256 maxUsdc = _min(usdcLeft, allowanceLeft);
+            if (maxUsdc > 0) {
+                _executeOrder(orders[i], maxUsdc);
+                lastPrice = orders[i].price;
+            } else {
+                break;
+            }
+        }
+    }
+
+    /// @dev Executes a single order, possibly partially. If the specified
+    ///      tender address does not exist, returns false. Otherwise
+    ///      executes the order and returns true in case of success.
+    /// @param order The order to execute
+    /// @param maxUsdc Maximum amount of USDC that can be spent on execution
+    /// @return Whether the order has been executed
+    function _executeOrder(Order memory order, uint256 maxUsdc)
+        internal
+        returns (bool)
+    {
+        if (!registry.isTenderAddress(order.redeemerTenderAddress)) {
+            // Only tender addresses are allowed
+            return false;
+        }
+        uint256 redeemerBlendBalance =
+                blend.balanceOf(order.redeemerTenderAddress);
+        uint256 blendAmount = _min(order.amount, redeemerBlendBalance);
+        uint256 usdcAmount = _blendToUsdc(blendAmount, order.price);
+        if (usdcAmount > maxUsdc) {
+            usdcAmount = maxUsdc;
+            blendAmount = _usdcToBlend(usdcAmount, order.price);
```

Comment:
I do not like this piece. Effectually it says that if all amount of USDC left is `10` USDC and price per blend in the order is `9 USDC`, we'll charge blend holder `2 BLND` (valued at `18 USDC`). True, we can expect that at most one BLEND will be charged extra, but still that's not good behaviour.

Case with roundup in `_blendToUsdc` is fine as only small fixed quantity can be extra charged, while with blend price possibly unbounded this is more of a problem.

I propose to replace these lines with:
```
uint256 blendAmount = _min(_min(order.amount, redeemerBlendBalance), maxUsdc.mul(PRICE_MULTIPLIER) `div` order.price);
uint256 usdcAmount = _blendToUsdc(blendAmount, order.price);

if (usdcAmount > maxUsdc) {
blendAmount = blendAmount - 1
usdcAmount = _blendToUsdc(blendAmount, order.price)
}
```

Reply:
> True, we can expect that at most one BLEND will be charged extra, but still that's not good behaviour

You meant 1 * 10^-18 BLEND

> I propose to replace these lines with

`if (usdcAmount > maxUsdc)...` lines handle USDC deficiency. If I understand your code correctly, you propose to ignore USDC balance, and in case of insufficient USDC use blend amount for calculations. How does subtracting 1 from blend amount help in case we have `1 USDC` left, and `blendAmount * price` is much larger?

Reply:
Notice `maxUsdc.mul(PRICE_MULTIPLIER) `div` order.price` in `_min` on the first line.

But ok, I think `1 * 10^-18 BLEND` is fine for an error.
I forgot we have this `10^-18` multiplier

Reply:
Ah, didn't notice that you've modified the first line. Seems legit